File: backend/src/main/java/lan/dk/podcastserver/manager/worker/tf1replay/TF1ReplayFinder.java
package lan.dk.podcastserver.manager.worker.tf1replay;

import com.github.davinkevin.podcastserver.service.HtmlService;
import com.github.davinkevin.podcastserver.service.ImageService;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.manager.worker.Finder;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.nodes.Document;
import org.springframework.stereotype.Service;

import javax.validation.constraints.NotEmpty;
import java.util.function.Function;
import java.util.regex.Pattern;

import static com.github.davinkevin.podcastserver.utils.MatcherExtractor.PatternExtractor;
import static com.github.davinkevin.podcastserver.utils.MatcherExtractor.from;
import static io.vavr.API.*;

/**
 * Created by kevin on 20/07/2016.
 */
@Service
@RequiredArgsConstructor
public class TF1ReplayFinder implements Finder {

    private static final PatternExtractor PICTURE_EXTRACTOR = from(Pattern.compile("url\\(([^)]+)\\).*"));

    private final HtmlService htmlService;
    private final ImageService imageService;

    @Override
    public Podcast find(String url) {
        return htmlService.get(url)
                .map(htmlToPodcast(url))
                .getOrElse(Podcast.DEFAULT_PODCAST);
    }

    private Function<Document, Podcast> htmlToPodcast(String url) {
        return p -> Podcast.builder()
                        .title(p.select("meta[property=og:title]").attr("content"))
                        .description(p.select("meta[property=og:description]").attr("content"))
                        .url(url)
                        .cover(getCover(p))
                        .type("TF1Replay")
                    .build();
    }

    private Cover getCover(Document p) {
        return PICTURE_EXTRACTOR.on(p.select(".focalImg style").html()).group(1)
                .orElse(Option(p.select("meta[property=og:image]").attr("content")))
                .map(this::getUrl)
                .map(imageService::getCoverFromURL)
                .getOrElse(Cover.DEFAULT_COVER);
    }

    private String getUrl(String url) {
        return Match(url).of(
                Case($(u -> u.startsWith("//")), u -> "http:" + u),
                Case($(), Function.identity())
        );
    }


    @Override
    public Integer compatibility(@NotEmpty String url) {
        return StringUtils.contains(url, "www.tf1.fr") ? 1 : Integer.MAX_VALUE;
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/tf1replay/TF1ReplayFinder.kt
package com.github.davinkevin.podcastserver.manager.worker.tf1replay

import arrow.core.getOrElse
import arrow.core.orElse
import arrow.core.toOption
import com.github.davinkevin.podcastserver.service.HtmlService
import com.github.davinkevin.podcastserver.service.ImageService
import com.github.davinkevin.podcastserver.utils.MatcherExtractor.Companion.from
import com.github.davinkevin.podcastserver.utils.k
import com.github.davinkevin.podcastserver.entity.Cover
import lan.dk.podcastserver.entity.Podcast
import com.github.davinkevin.podcastserver.manager.worker.Finder
import org.apache.commons.lang3.StringUtils
import org.jsoup.nodes.Document
import org.springframework.stereotype.Service

/**
 * Created by kevin on 20/07/2016.
 */
@Service
class TF1ReplayFinder(val htmlService: HtmlService, val imageService: ImageService) : Finder {

    override fun find(url: String): Podcast {
        return htmlService.get(url).k()
                .map { toPodcast(it, url) }
                .getOrElse { Podcast.DEFAULT_PODCAST }
    }

    private fun toPodcast(d: Document, anUrl: String) =
            Podcast().apply {
                title = d.select("meta[property=og:title]").attr("content")
                description = d.select("meta[property=og:description]").attr("content")
                url = anUrl
                cover = getCover(d)
                type = "TF1Replay"

            }

    private fun getCover(p: Document) =
            PICTURE_EXTRACTOR.on(p.select(".focalImg style").html()).group(1).k()
                    .orElse { p.select("meta[property=og:image]").attr("content").toOption() }
                    .map { toUrl(it) }
                    .flatMap { imageService.getCoverFromURL(it).toOption() }
                    .getOrElse { Cover.DEFAULT_COVER }

    private fun toUrl(url: String): String = when {
        url.startsWith("//") -> "http:$url"
        else -> url
    }


    override fun compatibility(url: String?) =
            if (StringUtils.contains(url, "www.tf1.fr")) 1
            else Integer.MAX_VALUE

    companion object {
        private val PICTURE_EXTRACTOR = from("url\\(([^)]+)\\).*")
    }
}
